numba/targets/rangeobj.py

```diff
@@ -65,6 +65,25 @@ def range3_impl(context, builder, sig, args):
                                   range_state_type,
                                   state._getvalue())
 
+    @lower_builtin(len, range_state_type)
+    def range_len(context, builder, sig, args):
+        """
+        len(range)
+        """
+        (value,) = args
+        state = RangeState(context, builder, value)
+        res = RangeIter.from_range_state(context, builder, state)
+        return impl_ret_untracked(context, builder, int_type, builder.load(res.count))
+
+    @lower_builtin(len, range_iter_type)
+    def range_iter_len(context, builder, sig, args):
```

Comment:
Yes, you are right that `len` over range iterator is not supported in Python.

But the purpose of this code is for internal use. Python comprehension compiles into a closure call to an inner function, and in this call, only an iterator object is passed instead of the range object. I'm implementing a conversion from list comprehension to array comprehension for Numpy call `array([... for ...])`. In this conversion, I need to allocate the array before loop starts, but the list size is only available in the iterator. So things would be a lot easier if I could produce Numba IR that calls `len` on the iterator.

I did consider a few alternatives without adding the above code:

1. Allocate the array before the closure call, pass it as an extra argument to the actual all. Now the question is how to get the type of the element before the closure call, and so on.

2. Pass the actual range object as an extra argument to the closure call.

It is just extra trouble in implementing either approach for the benefit of getting hold of an array size, which is already avaliable in the range iterator object. So they don't feel as clean as supporting calling `len` directly over range iterator object from within Numba IR.